### ISSUE TITLE
Support bool mask for logits_to_keep and expose last_hidden_state in Gemma4ForConditionalGeneration

### ISSUE THREAD

**Issue by jp1924** [huggingface/transformers#48784]
### Feature request

Two changes to `Gemma4ForConditionalGeneration.forward`:
1. Accept a bool tensor for `logits_to_keep` to support non-contiguous masking.
2. Add a flag independent of `output_hidden_states` that returns `last_hidden_state`.

### Motivation

SFT training with sequence packing concatenates multiple examples into one sequence, and some spans within that sequence need to be excluded from logits computation. These spans are non-contiguous, so a slice or an int can't express them. Right now this requires monkey-patching `forward()` to slice `hidden_states` directly.

`last_hidden_state` is unexposed for the same reason. Getting the final layer's hidden state for anything other than logits currently requires `output_hidden_states=True`, which stores every layer's hidden state, then indexing `hidden_states[-1]`. This keeps unneeded intermediate layers in memory, and `forward()` still doesn't return `last_hidden_state` directly, so this also needs a monkey patch.

Both problems come down to the same thing: neither is solvable without a monkey patch.

### Your contribution

Willing to submit a PR implementing this for `Gemma4ForConditionalGeneration`.

**Comment by VimalN2005** [comment]
Hi @jp1924, completely agree with this! For sequence packing during SFT, non-contiguous masking via boolean tensor is essential to avoid unnecessary monkey-patching.

Exposing `last_hidden_state` directly without forcing `output_hidden_states=True` will also save significant GPU VRAM by not caching all intermediate layer representations. Looking forward to your PR!

**Comment by jp1924** [comment]
@zucchini-nlp @ArthurZucker
Opened a draft PR (#48825) implementing both halves of this issue
1. Future models will likely need more varied ways to select which `last_hidden_state` positions matter, and slice-based `logits_to_keep` only expresses a single contiguous range, so it can't cover that. In the draft, supporting both a bool mask and a slice turned out to be two genuinely different indexing operations that don't compose without branching. Instead of keeping slice support and adding a mask on top, would it make more sense to move `logits_to_keep` itself to a tensor-based mask?
2. `logits_to_keep` isn't Gemma4-specific, most decoder models carry the same parameter. If (1) is the right direction, this isn't a single-model change, it applies across the codebase. transformers tends to value consistent patterns across models, so I'm unsure whether it's fine to change this shared pattern in Gemma4 first. Is starting here and extending to other models once the direction is confirmed acceptable, or should the design be settled centrally before this PR moves forward?
3. This follows from (2). If the same change needs to apply beyond Gemma4, the scope grows fast: of the 73 `ForConditionalGeneration` models that take `logits_to_keep`, 69 have no `last_hidden_state` field on their output at all. Landing this everywhere in one PR isn't realistic. If it does need to land everywhere eventually, is it fine to land it in Gemma4 first and open follow-up PRs for the rest?

**Comment by jp1924** [comment]
https://github.com/huggingface/transformers/pull/48811#pullrequestreview-5209553779
> nope, not needed imo and we have been discussing several times with vision team a feat to pass layer indices to `output_hidden_states`
> So imo a better solution is to work towards that, which will be handled internally soon (hopefully)

Agreed, letting `output_hidden_states` accept layer indices is a cleaner direction than a separate flag for this. Same tradeoff applies to #48784, will fold this into the discussion there.

**Comment by zucchini-nlp** [comment]
>  select which last_hidden_state positions matter

Fyi t clear doubts, it will return a whole tensor still so kinda not the same as requested here. Btw, since you have been contrib a lot, i think you can work on allowing `indices` for the output hidden states (if you wish ofc)
Prob we can support it easily via `can_record_output` decorator and old models without a decorator will not support it. We have the decorator in all important and recent models anyway. As long as the updates are minimal and compltely BC, it'll work for us

**Comment by zucchini-nlp** [comment]
>  some spans within that sequence need to be excluded from logits computation

oke, coming back to read properly. Do we need to not return logits in those spans or just not compute loss over it (e.g. image placeholder tokens don't contrib to loss ideally)? If it is ony for loss, usualy we handle that via setting `-100` in `labels`, if not lmk the use-case pls

**Comment by jp1924** [comment]
@zucchini-nlp 
Ah, my earlier explanation was off, sorry for the confusion.

The mask I'm proposing only applies to the hidden_state fed into `lm_head`. `last_hidden_state` as returned in `ModelOutput` stays unmasked, the full original tensor. Once a caller has `last_hidden_state`, they can mask it however they want on their end.

Touching logits isn't a loss problem, it's a memory optimization.

> image placeholder tokens don't contrib to loss

Right, and that's exactly the case where `-100` already works fine. The intent behind moving `logits_to_keep` to a mask isn't about excluding something from loss, it's resource optimization: positions masked with `-100` are excluded from the loss, but the logits tensor is still `(bsz, seq, vocab)` regardless, so it still takes up that much memory even for positions that never contribute to loss. Masking the hidden_state before the vocab projection, e.g. skipping the prompt span in packed SFT, avoids materializing that tensor for positions that don't need it. That's why the mask needs to support non-contiguous spans.

**Comment by zucchini-nlp** [comment]
Ahh right right, the dummy placeholders can be really huge in some models, valid point! Personally that does sound like a viable solution to me. Re the 69 with no last_hidden_state field, i suppose that doesn't matter since we return `logits` of certain sliced shape and afaik no model returns last hidden state from a `Generative` class

cc @vasqu , since we'll need at least one core person's opinion re `logits_to_keep` part - it will touch a lot of models

**Comment by jp1924** [comment]
> Prob we can support it easily via can_record_output decorator and old models without a decorator will not support it. We have the decorator in all important and recent models anyway. As long as the updates are minimal and compltely BC, it'll work for us

Interested in the output_hidden_states-with-layer-indices idea. 
It's 11pm here right now, so I can't give this a proper look tonight. I'll review it and get back tomorrow.

Also, if I do end up picking this up, Sep 21-25 is a national holiday in Korea (Chuseok), so work or replies from me could be slow by about a week during that stretch. Not dropping it, just flagging the delay in advance.

**Comment by zucchini-nlp** [comment]
dw, no rush at all. That idea has been going back and forth for a while and is not super high prio, since it is only an optim on existing stuff. LMK when you need any help/brainstorming/review and happy Chuseok 🤗

**Comment by vasqu** [comment]
just chiming in to glance over the quick rundown
1. Return last hidden state before logits (on ForXXX) -> note that bites us on the speculative decoding as well so the impact radius might be quite nice to avoid expensive memory collection. I remember somewhere where we could already only select specific layers but yea not a bad idea
2. Logits to keep is a simple int like that only keeps the last hidden states before computing the logits -> you want to reduce the unnecessary memory overhead correct? Yep, I think it's a good idea to give more control here. I havent thought about that but does that happen a lot in SFT? I'd expect that most should be valid tokens no? Could I have a dummy example where this would make sense -> maybe some RL type of prefix tuned training?

Anyways, open for 1. and for 2. a bit reluctant but not opposed (need to see the usecase in more detail than SFT). Also, no worries we are in no rush here, but these could be nice optims indeed!

**Comment by zucchini-nlp** [comment]
> Return last hidden state before logits 

nah iiuc shoudnt be needed anymore if we output hidden states by layer id, to not return all layers. Kinda redundant to return last hidden as new field

**Comment by vasqu** [comment]
No I didnt mean as an additional field, but instead of collecting all layer hidden states -> return only the last (so essentially what you describe by last id layer)

**Comment by vasqu** [comment]
https://github.com/huggingface/transformers/blob/3f162ca5fd2daef9cb36841be1337ad657addbfe/src/transformers/generation/candidate_generator.py#L1601

E.g. here we do collect all hidden states and then filter by what we need - if we could just give direct ids to what we need == less memory

**Comment by zucchini-nlp** [comment]
yep, exactly what we've been discussin above in linked PR 😅 
too many linked things, that is defi on the radar and helpful in many other cases with mutimodals

**Comment by jp1924** [comment]
@vasqu @zucchini-nlp 

> you want to reduce the unnecessary memory overhead correct?

yep

> I havent thought about that but does that happen a lot in SFT? I'd expect that most should be valid tokens no? Could I have a dummy example where this would make sense -> maybe some RL type of prefix tuned training?

TRL issues [#3549](https://github.com/huggingface/trl/issues/3549) and [#2469](https://github.com/huggingface/trl/issues/2469) are examples, this comes up in GRPO, DPO, and RLOO too. It's not SFT-specific, it shows up in enough different cases that pinning it down to one single exact use case is hard.

> Prob we can support it easily via can_record_output decorator and old models without a decorator will not support it. We have the decorator in all important and recent models anyway. As long as the updates are minimal and compltely BC, it'll work for us



I'll take on the output_hidden_states layer-indices work.
Turns out most of it already landed in #48081, output_hidden_states already accepts a list of layer indices and only materializes those, the rest come back as None.
Planning to build on top of that: keep the default False, True still returns everything, a list of indices returns just those layers, and -1 returns the last layer's hidden_states.

Also going to make logits_to_keep accept a 2D mask instead of just a slice.

**Comment by vasqu** [comment]
Ok makes sense just for terminology it's RL and not SFT, no? At least it makes more sense to me then (it's been a while that ive been on the training side sorry!)

Sounds good (and dang I knew there was something already :D)

**Comment by jp1924** [comment]
@vasqu @zucchini-nlp

Well, scope creep, as always.

Switching `logits_to_keep` to a 2D mask means `generate()` needs work too.
I didn't get why `logits_to_keep` existed in the first place, but reading the `generate()` code made it click.
It's used to compute logits only for the last position needed for the next-token prediction during the prefill step, since computing logits over the whole prompt there would be wasteful. That's why a slice was always enough.

So here's how the scope of this work expands:
this ends up being about taking a feature that was originally built only for `generate()`, and widening it so training (non-contiguous position selection in packed sequences) can use it too.

If I don't touch `generate()` and only fix `forward`, I could just support both slice and mask there.
But that's not really solving the underlying problem, so it doesn't sit right with me.

I'll go with whatever you all decide, either direction works for me.

- Former: add slice/2D-mask support only to the `ForXXX` models' `forward`, within a scope we can actually review.
- Latter: make `logits_to_keep` consistently 2D-mask across both `generate()` and `forward`. This means touching every model's `forward` that's wired to `logits_to_keep`.

If it's the latter, `generation/utils.py` and `generation/continuous_batching/model_runner.py` both need changes, and here's what I've found needs touching:

- `generation/utils.py`:
  - Where `generate()` sets `model_kwargs["logits_to_keep"] = 1` near the start. Needs to build a mask instead of an int.
  - Where assisted decoding overrides it with `model_inputs["logits_to_keep"] = candidate_length + 1`. Same, mask-based.
  - `outputs.logits[:, -1]` in the greedy/sample loop, `model_outputs.logits[:, -1, :]` in the beam search loop, and `outputs.logits[:, -candidate_length - 1:]` in assisted decoding. All three assume a 3D `(batch, kept, vocab)` shape, so if this goes mask-based, that shape assumption needs fixing too.
- `generation/continuous_batching/model_runner.py`:
  - Where `batch_data["logits_to_keep"] = batch_data["logits_indices"]` gets set. `logits_indices` is a 1D tensor collecting just the last token position of each request after concatenating several different requests into one flat sequence. It's the same shape as the non-contiguous selection problem we're discussing right now, and it turns out the 1D-tensor form is already being used for exactly that.
  - `_forward_process_and_sample` and `_sample` also assume a 3D `(batch, kept, vocab)` shape, so those need fixing too.

If it's the latter, I'll write up a comment mentioning `@Cyrilvallez` explaining where things stand. Nothing's decided yet, so this is just a heads-up mention for now.

This feels a bit like that ad Shackleton supposedly ran before his Antarctic expedition.

**Comment by zucchini-nlp** [comment]
I think we will need a deprecation cycle in any case, since `logits_to_keep` might be used by remote code in unexpected way, i know a few models that even saved `config.logits_to_keep` and pass it over to `generate`. We can always deprecate and workaround those cases by force-converting int to a tensor in the end
(i realize that is a really bad pattern that should've never been merged, it is what it is though 🫠 )

For now, I'd say we go with former option to keep BC and the git diff manageable for review first -> we merge it hopefully -> then we can think about moving into tensor-only format in the least breaking way

**Comment by jp1924** [comment]
@zucchini-nlp

Agreed.
To keep BC, I'll implement both slice and mask support for now.
One more thing, just need your opinion, should this apply to gemma-4 only, or to every other model that uses logits_to_keep too?

**Comment by vasqu** [comment]
It's always a bad idea to override something that has been used in a certain way. The better way is to support both variation
1. A normal int as previously
2. A mask that is index based (so a bool mask that is true on the correct positions)

I'm not sure why we only try to support one way 🤔 generate can then keep the simple int based version and power users can opt for a more sophisticated selection strategy

**Comment by vasqu** [comment]
> One more thing, just need your opinion, should this apply to gemma-4 only, or to every other model that uses logits_to_keep too?

Tbh it goes beyond gemma so all models 😅

**Comment by jp1924** [comment]
@zucchini-nlp @vasqu 
Tried to keep the code minimal while avoiding anything that could trip up torchdynamo/compile, and to land on something readable.

###  Single expression (tuple + Ellipsis), one `lm_head` call:
```python
    slice_indices = (
        logits_to_keep
        if isinstance(logits_to_keep, torch.Tensor) and logits_to_keep.dtype == torch.bool
        else (
            slice(None),
            slice(-logits_to_keep, None) if isinstance(logits_to_keep, int) else logits_to_keep,
            ...,
        )
    )
    logits = self.lm_head(hidden_states[slice_indices])
```

Shrinks to one line, but depending on the condition, `slice_indices` ends up holding either a plain tensor or a 3-tuple under the same name, which made it awkward to read.

### try/except to drop `isinstance` entirely, EAFP style:
```python
    try:
        indexed = hidden_states[:, slice(-logits_to_keep, None), :]
    except TypeError:
        indexed = hidden_states[:, logits_to_keep, :]
    except RuntimeError:
        indexed = hidden_states[logits_to_keep]
    logits = self.lm_head(indexed)
```

Drops `isinstance`, but exception-based branching risks breaking the compile graph, so shelved it.

### Ended up settling on this:

```python
    if isinstance(logits_to_keep, torch.Tensor) and logits_to_keep.dtype == torch.bool:
        logits = self.lm_head(hidden_states[logits_to_keep])
    else:
        slice_indices = slice(-logits_to_keep, None) if isinstance(logits_to_keep, int) else logits_to_keep
        logits = self.lm_head(hidden_states[:, slice_indices, :])
```

Let me know if there's feedback.
Once this form is settled, I'll roll it out to the rest of the models.

Out of 257 classes total (232 model directories), 240 already have this exact pattern in their forward, so those are straightforward to patch.

One exception is `ShieldGemma2ForImageClassification`, which takes the inner model's output and indexes it again itself with `logits[:, -1, [yes_idx, no_idx]]`.
That assumes a 3D output, so if a bool mask comes in and flattens it to 2D, this breaks. Not patching that one in this PR.

**Comment by zucchini-nlp** [comment]
Wait a sec, so we support non-int LongTensors atm 😮 which means we'll end up with supported int or LongTensor of `size<seqlen`  or BoolTensor of input shape. That is quite a lot of semantics baked in single arg imo, though it's the only way to get memory optim for packed training... 
We just need to document it well, since the `out.logits` will have different sizes when sliced (`bs*packed_len`, `dim`)

For easier code inspection, i prefer splitting things into separate lines - list comprehension works well only for shorter seqs 


> One exception is ShieldGemma2ForImageClassification

That is a classifier model with 2 classes, so i don't think it even needs to slice into logits no?

**Comment by vasqu** [comment]
I like the last option but a bit simplified
```py
    if isinstance(logits_to_keep, torch.Tensor):
        hidden_states = hidden_states[logits_to_keep])
    elif isinstance(logits_to_keep, int):
        slice_indices = slice(-logits_to_keep, None)
        hidden_states = hidden_states[:, slice_indices, :]

    logits = self.lm_head(hidden_states)
```

that assumes that a tensor would always be bool which imo should be a fair assumption (esp if we also propery type this)

**Comment by jp1924** [comment]
@zucchini-nlp

> We just need to document it well, since the out.logits will have different sizes when sliced (bs*packed_len, dim)
> For easier code inspection, i prefer splitting things into separate lines - list comprehension works well only for shorter seqs

Ok, will address both.

> That is a classifier model with 2 classes, so i don't think it even needs to slice into logits no?

Ah, that's on me for how I framed it. 
`ShieldGemma2` was never actually in scope to begin with. 
We're only applying this to `ForCausalLM`, `ForConditionalGeneration`, and `LMHead-equivalent classes` (`LMHeadModel`, `BertGenerationDecoder`, `ReformerModelWithLMHead`, etc.), classification/image-classification heads are excluded entirely. 
Can you confirm that's the right scope, or is there something else that should be in/out?

@vasqu

So are you saying `logits_to_keep` should be restricted up front to only int or a `2D BoolTensor`, guarded/erroring on anything else? 
If a `1D IndexTensor` comes in, it can misbehave without raising (e.g. `torch.tensor([-1])` silently hits the batch dim regardless of batch size, or `torch.tensor([0, 1])` with `batch_size=4` silently returns batch rows 0 and 1 in full instead of slicing the sequence dimension).

`logits_to_keep` was originally designed to accept both an int and a 1D tensor, so we went with this instead:

```python
if isinstance(logits_to_keep, int):
    slice_indices = slice(-logits_to_keep, None)
    hidden_states = hidden_states[:, slice_indices, :]
elif logits_to_keep.dim() == 2:
    hidden_states = hidden_states[logits_to_keep]
else:
    hidden_states = hidden_states[:, logits_to_keep, :]

logits = self.lm_head(hidden_states)
```

What do you think?

**Comment by zucchini-nlp** [comment]
> Can you confirm that's the right scope, or is there something else that should be in/out?

yep, only generative models with a head!


> elif logits_to_keep.dim() == 2:
    hidden_states = hidden_states[logits_to_keep]
else:
    hidden_states = hidden_states[:, logits_to_keep, :]

This is still weird imo, unless we check the dtype - thus my comment before on `LongTensor` support. It's still power feature, so i doubt many ppl use it but better be safe. Your last code sniippet option from prev comment is the one needed - accepted options are legacy "int or 1d long tensor" and the new "packed 2d boolean tensor"

**Comment by jp1924** [comment]
@zucchini-nlp
Ah got it, makes sense
dtype is the actual signal, not shape. Going with dtype then:

```python
if isinstance(logits_to_keep, int):
    slice_indices = slice(-logits_to_keep, None)
    hidden_states = hidden_states[:, slice_indices, :]
elif logits_to_keep.dtype == torch.bool:
    hidden_states = hidden_states[logits_to_keep]
else:
    hidden_states = hidden_states[:, logits_to_keep, :]

logits = self.lm_head(hidden_states)
```

Same as what I originally settled on, just merged into a single lm_head call.
